### PR TITLE
Add multiple source links for teleporters

### PR DESCRIPTION
**Pull request by Th30n**

### CODE REVIEW

**Comment by Th30n** [comment]
@johnnovak this my initial implementation for the "teleport destinations with multiple sources" feature. Feel free to suggest a different approach or tweaking. Otherwise, when/if the mechanics are good to go, I'll update the manual, changelog, etc.

**Comment by johnnovak** [comment]
This is absolutely outstanding work, @Th30n! Sorry for the delayed respone, but my 95 year old father died and I need to support my mother during this period and organise a lot of things... I'll be back in business around 20 Oct, then I'll review and test this properly.

**Comment by Th30n** [comment]
> This is absolutely outstanding work, @Th30n! Sorry for the delayed response, but my 95 year old father had died and I need to support my mother during this period and organise a lot of things... I'll be back in business around 20 Oct, then I'll review and test this properly.

I am sorry to hear that! That is some nice age under you father's belt, I hope he had a good life. My sincere condolences to you and your family.

File: src/persistence.nim [hunk 1]
```diff
@@ -290,7 +290,7 @@ proc readLinks_v1(rr; levels: seq[Level]): Links =
   debug(fmt"  numLinks: {numLinks}")
   checkValueRange(numLinks, "links.numLinks", NumLinksLimits)
 
-  result = initBiTable[Location, Location](nextPowerOfTwo(numLinks))
+  result = initLinks(nextPowerOfTwo(numLinks))
```
Comment (johnnovak) [inline src/persistence.nim:293]: I would only pre-init for performance intensive code where the re-allocations do actually cause a measurable/observable difference. Otherwise it just adds noise to the code, so unless you can prove this makes a big difference, can you please remove it?

File: src/map.nim [hunk 2]
```diff
@@ -253,9 +256,10 @@ proc normaliseLinkedStairs*(m; level: Natural) =
 
       if f in LinkStairs:
         let this = Location(level: level, row: r, col: c)
-        let that = m.getLinkedLocation(this)
-        if that.isSome:
-          let that = that.get
+        let that = m.getLinkedLocations(this)
+        if that.len > 0:
+          assert that.len == 1, "Stairs should not have linked multiple locations"
```
Comment (johnnovak) [inline src/map.nim:261]: Nice defensive style 👍🏻

File: src/common.nim [hunk 3]
```diff
@@ -86,7 +86,11 @@ type
     sortedLevelIdxToLevelIdx*: Table[Natural, Natural]
 
 
-  Links* = BiTable[Location, Location]
+  Links* = object
+    # TODO: Maybe use a fixed size array instead of HashSet?
```
Comment (johnnovak) [inline src/common.nim:90]: HashSet is fine as it's not performance intensive.

File: src/main.nim [hunk 4]
```diff
@@ -274,6 +275,10 @@ type
 
     linkSrcLocation:    Location
 
+    jumpToSrcLocation:   Location
+    jumpToDestinations:  seq[Location]
+    jumpToDestinationIx: Natural
```
Comment (johnnovak) [inline src/main.nim:280]: `jumpToDestLocations` and `jumpToDestLocationIdx` for consistency.

File: src/main.nim [hunk 5]
```diff
@@ -319,7 +324,8 @@ type
     emSelectDraw,
     emSelectErase,
     emSelectRect,
-    emSetCellLink
+    emSetCellLink,
+    emSelectJumpToLink
```
Comment (johnnovak) [inline src/main.nim:328]: `emSelectJumpToLinkSrc`

File: src/main.nim [hunk 6]
```diff
@@ -5480,6 +5489,15 @@ proc handleLevelMouseEvents(a) =
 
     else: discard
 
+# }}}
+
+# {{{ setSelectJumptoLinkActionMessage()
+proc setSelectJumptoLinkActionMessage(a) =
```
Comment (johnnovak) [inline src/main.nim:5495]: `setSelectJumpToLinkSrcActionMessage`

File: src/main.nim [hunk 7]
```diff
@@ -5480,6 +5489,15 @@ proc handleLevelMouseEvents(a) =
...
+  let currIx = a.ui.jumpToDestinationIx + 1
+  let count = a.ui.jumpToDestinations.len
+  setStatusMessage(NoIcon, fmt"Select destination ({currIx} of {count})",
```
Comment (johnnovak) [inline src/main.nim:5498]: So we're selecting between multiple source cells here, not destinations. Also, it would be nice to include the type of the link in the message like in other places.

```nim
fmt"Select {linkFloorToString(floor)} source ({currIx} of {count})"
```

File: src/main.nim [hunk 8]
```diff
@@ -5480,6 +5489,15 @@ proc handleLevelMouseEvents(a) =
...
+                   @[IconArrowsAll, "next/prev", "Enter", "confirm", "Esc", "cancel"], a)
```
Comment (johnnovak) [inline src/main.nim:5499]: There is no difference in behaviour when pressing "Enter" or "Esc" -- both just exit the jump to link mode. Therefore, this should be `@[IconArrowsAll, "next/prev", "Enter/Esc", "exit"]`.

File: src/main.nim [hunk 9]
```diff
@@ -5858,9 +5876,17 @@ proc handleGlobalKeyEvents(a) =
...
       elif ke.isShortcutDown(scJumpToLinkedCell):
-        let otherLoc = map.getLinkedLocation(cur)
-        if otherLoc.isSome:
-          moveCursorTo(otherLoc.get, a)
+        let otherLocs = map.getLinkedLocations(cur)
+        if otherLocs.len == 1:
+          moveCursorTo(otherLocs.first.get, a)
+        elif otherLocs.len > 1:
+          ui.jumpToSrcLocation = cur
+          ui.jumpToDestinations = otherLocs.toSeq
```
Comment (johnnovak) [inline src/main.nim:5884]: Yeah, so with src and dest swapped in the names this code would make a lot more sense (we support single destinations and multiple sources).

File: src/main.nim [hunk 10]
```diff
@@ -6485,6 +6511,31 @@ proc handleGlobalKeyEvents(a) =
...
+
+      handleMoveKeys(ke, allowWasdKeys=true, allowRepeat=false, handleMoveKey)
+
+      if ke.isShortcutDown(scAccept):
+        ui.editMode = emNormal
+        clearStatusMessage(a)
+      elif ke.isShortcutDown(scCancel):
+        moveCursorTo(ui.jumpToSrcLocation, a)
+        ui.editMode = emNormal
+        clearStatusMessage(a)
```
Comment (johnnovak) [inline src/main.nim:6536]: This is nice overall, but as I mentioned above, I'd like a different behaviour here that makes the feature more consistent with the single src & dest link case where you can repeatedly press `G` to jump back and forth between the src and dest cells.

We should store the last selected source cell, and pressing `G` in link source select mode should still jump back and forth between the fixed destination cell and the last selected source cell.

File: src/links.nim [hunk 11]
```diff
@@ -1,21 +1,42 @@
...
 
-import bitable
 import common
 import rect
 import selection
 import utils
 
-export bitable
-
+const DefaultInitialSize = 64
```
Comment (johnnovak) [inline src/links.nim:10]: This is not performance sensitive code, so let's ditch the initial size setting.

File: src/links.nim [hunk 12]
```diff
@@ -177,4 +227,84 @@ proc shiftLinksInLevel*(l; level: Natural, rowOffs, colOffs: int,
...
+  let loc1 = Location(level: 0, row: 0, col: 0)
+  let loc2 = Location(level: 0, row: 1, col: 0)
+  let loc3 = Location(level: 0, row: 0, col: 1)
+  var l = initLinks()
+
+  assert l.len == 0
+  assert l.hasWithSrc(loc1) == false
+  assert l.hasWithSrc(loc2) == false
+  assert l.hasWithDest(loc1) == false
+  assert l.hasWithDest(loc2) == false
```
Comment (johnnovak) [inline src/links.nim:241]: minor: maybe do `assert not ...` instead

File: src/persistence.nim [hunk 13]
```diff
@@ -290,7 +290,7 @@ proc readLinks_v1(rr; levels: seq[Level]): Links =
...
```
Comment (johnnovak) [inline src/persistence.nim:294]: Also, you should update `CurrentMapVersion` on line 26 to `2.`

**Comment by Th30n** [comment]
Thanks for the review, I'll work on this over the weekend

**Comment by johnnovak** [comment]
> Thanks for the review, I'll work on this over the weekend

Sounds great @Th30n, no rush. May I suggest to address issues in separate commits (as opposed to force pushing) to make the review easier? Also, it's best to make every single commit compilable for easy bisecting (if needed). 

I've upgraded to Nim 1.6.8, I think you should do the same (there will be some naming related warnings; just ignore those for now).

**Comment by Th30n** [comment]
Will do that.

Btw, I only force push after rebasing (on master in this case), and I only rebase to keep the commit history linear so as to make `git log` and `git bisect` even easier to work with.

**Comment by Th30n** [comment]
> I think we should allow multiple sources for all link types.

I've been thinking about this, and now I'm not so sure about it. I think it makes sense for one way mechanisms, like teleporters typically are. I guess it would then also make sense for pits since you typically enter on one end cannot return at the exit point.
But what about stairs and doors? They are typically bidirectional, and games usually don't allow you to select the final destination i.e. they utilize single source & single destination in bidirectional links.

The only example I can think of that breaks this simplicity is Might & Magic via its teleportation mechanism in towns. Each town has an area with a teleportation transport service where the player can pick (and pay) to be teleported to another town of choice.

In summary, I think we should postpone the decision to allow other link types having multiple sources for now.

> I'm also thinking it could be handy to show some extra visual marker in multi-source destination cells, but I'm not sure how... I'd need to experiment with this a bit, and perhaps it's outside of the scope of the current PR anyway.

I agree, I think it would also be great to somehow highlight all the sources on the current floor when hovering over a destination.

File: src/main.nim [hunk 14]
```diff
@@ -5480,12 +5480,15 @@ proc handleLevelMouseEvents(a) =
...
 
-# {{{ setSelectJumptoLinkActionMessage()
-proc setSelectJumptoLinkActionMessage(a) =
-  let currIx = a.ui.jumpToDestinationIx + 1
-  let count = a.ui.jumpToDestinations.len
-  setStatusMessage(NoIcon, fmt"Select destination ({currIx} of {count})",
-                   @[IconArrowsAll, "next/prev", "Enter", "confirm", "Esc", "cancel"], a)
+# {{{ setSelectJumpToLinkActionMessage()
+proc setSelectJumpToLinkSrcActionMessage(a) =
+  let currIx = a.ui.jumpToSrcLocationIdx + 1
```
Comment (johnnovak) [inline src/main.nim:5486]: `currIdx`

File: src/main.nim [hunk 15]
```diff
@@ -275,9 +275,9 @@ type
 
     linkSrcLocation:    Location
 
-    jumpToSrcLocation:   Location
-    jumpToDestinations:  seq[Location]
-    jumpToDestinationIx: Natural
+    jumpToDestLocation:   Location
+    jumpToSrcLocation:    seq[Location]
```
Comment (johnnovak) [inline src/main.nim]: `jumpToSrcLocations` (should be plural)

File: src/main.nim [hunk 16]
```diff
@@ -5870,13 +5871,28 @@ proc handleGlobalKeyEvents(a) =
...
-          ui.jumpToSrcLocationIdx = 0
-          moveCursorTo(ui.jumpToSrcLocation[ui.jumpToSrcLocationIdx], a)
+          if ui.jumpToDestLocation == cur:
+            # Last time we jumped to multiple sources we used this destination,
+            # so continue selecting sources from the source we left at.
+            let oldIdx = ui.jumpToSrcLocation.find(ui.lastJumpToSrcLocation)
+            if oldIdx == -1:
+              # The source we left at last time no longer exists (e.g. the user
+              # deleted it), so reset from beginning.
+              ui.jumpToSrcLocationIdx = 0
```
Comment (johnnovak) [inline src/main.nim:5888]: Nice that you thought of this case 👍🏻 It can easily happen with undo/redo as well.

**Comment by johnnovak** [comment]
Hey @Th30n, thanks for addressing my comments, this is coming together very nicely! 🎉 I noticed a few very minor remaining naming inconsistencies, plus two issues that affect functionality after doing some further testing.

Thanks for having a stab at updating the documentation too! I realised I haven't mentioned it anywhere that the doco is generated with Sphinx and lives in the https://github.com/johnnovak/gridmonger-site repo. I'll document that to make it clear, but don't worry about it, I'll update the doco based on your input after we merged this in. I have your current changes safely stashed away on a local branch, please feel free to remove the doco updates from the commit history in this repo.

Then I'll do the release, plus I'll automate it a bit more because right now it's a very manual and error prone process. But... don't worry about that either for now, I'll take care of it 😄 

> In summary, I think we should postpone the decision to allow other link types having multiple sources for now.

I agree with your analysis, it's best to postpone it, after all. 👍🏻 

> I think it would also be great to somehow highlight all the sources on the current floor when hovering over a destination.

Yep, I was thinking of something along these lines. This could be an enhancement further down the road and it requires experimentation, so let's not overload the current PR with this stuff. I suggest to focus on getting this first iteration of the feature in first.

So the two issues that need addressing:

### Issue 1 — Jumping back and forth with G for the first time (edge case)

1. Create a teleport source
2. Link it to a teleport destination
3. Create a second teleport source, and link it to the same destination
4. Place the cursor at the _second_ teleport source, and press `G` — the cursor jumps to the destination, this is OK
5. Now press `G` again — the cursor jumps to the _first_ teleport source, which is wrong; it should have jumped to the _second_ (the last location)

After this, when selecting between the teleport sources and pressing `G` repeatedly, the last source position  is  correctly remembered. The issue only happens when using the `G` repeatedly on a _newly created_ multi-source set of linked teleports, and using any other source than the first as the very first starting point.

### Issue 2 — Trail mode compatibility

When trail mode (`T` key) is on, and `G` is pressed on a destination cell that has multiple sources, the trail drawing should be temporarily disabled while in teleport source selection mode. When `Enter` or `Esc` is pressed, trail mode should be re-enabled (and therefore the trail will be drawn in the selected cell _only_ after exiting source select mode).

**Comment by johnnovak** [comment]
Thanks for adding this great feature @Th30n and addressing all the comments! Merged. I'll update the doco in the next few days and release a new version.
